File: JavaToKotlin/apps(kotlin)/src/main/kotlin/io/dustin/apps/board/api/usecase/follow/WriteFollowUseCase.java
package io.dustin.apps.board.api.usecase.follow;

import io.dustin.apps.board.domain.follow.model.Follow;
import io.dustin.apps.board.domain.follow.model.dto.FollowDto;
import io.dustin.apps.board.domain.follow.service.WriteFollowService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class WriteFollowUseCase {

    private final WriteFollowService writeFollowService;

    public FollowDto execute(Long followerId, Long followingId) {
        Follow follow = writeFollowService.create(followerId, followingId);
        FollowDto dto = FollowDto.from(follow);
        return FollowDto.from(follow);
    }
}

File: apps/src/main/kotlin/io/dustin/apps/board/api/usecase/follow/WriteFollowUseCase.kt
package io.dustin.apps.board.api.usecase.follow

import io.dustin.apps.board.domain.follow.model.dto.FollowDto
import io.dustin.apps.board.domain.follow.service.ReadFollowService
import io.dustin.apps.board.domain.follow.service.WriteFollowService
import io.dustin.apps.common.utils.duplicate


import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class WriteFollowUseCase (
    private val readFollowService: ReadFollowService,
    private val writeFollowService: WriteFollowService
) {

    @Transactional
    fun execute(followerId: Long, followingId: Long): FollowDto {
        val followOjb = readFollowService.getFollowObj(followerId, followingId)

        if(followOjb == null){
            val follow  = writeFollowService.create(followerId, followingId)
            return FollowDto.from(follow)
        }
        else{
            return throw duplicate("이미 팔로우한 유저입니다.")
        }

    }
}
